internal/git/git.go

```diff
@@ -0,0 +1,234 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package git
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/utils/merkletrie"
+	"go.uber.org/thriftrw/compile"
+	"go.uber.org/thriftrw/internal/compare"
+)
+
+// NewGitFS creates an implementation of FS to use git to discover
+// Thrift changes and previous version of a Thrift file.
+func NewGitFS(gitDir string, repo *git.Repository, from bool) *FS {
+	if repo == nil {
+		var err error
+		repo, err = git.PlainOpenWithOptions(gitDir, &git.PlainOpenOptions{
+			DetectDotGit:          true,
+			EnableDotGitCommonDir: true,
+		})
+		if err != nil {
+			return nil
+		}
+	}
+
+	return &FS{
+		repo:   repo,
+		gitDir: gitDir,
+		from:   from,
+	}
+}
+
+// Compare takes a path to a git repository and returns errors between HEAD and HEAD~
+// for any incompatible Thrift changes between the two shas.
+func Compare(path string) (compare.Pass, error) {
+	pass := compare.Pass{}
+	r, err := git.PlainOpenWithOptions(path, &git.PlainOpenOptions{
+		DetectDotGit:          true,
+		EnableDotGitCommonDir: true,
+	})
+	if err != nil {
+		return pass, err
+	}
+	fs := NewGitFS(path, r, false)
+	fsFrom := NewGitFS(path, r, true)
+
+	changed, err := findChangedThrift(path)
+	if err != nil {
+		return pass, err
+	}
+	var errs error
+	for _, c := range changed {
+		var toModule *compile.Module
+		if c.change == merkletrie.Modify {
+			toModule, err = compile.Compile(c.file, compile.Filesystem(fs))
+			if err != nil {
+				return pass, err
+			}
+		} else if c.change == merkletrie.Delete {
+			// something got deleted, so we are creating an empty module here.
+			toModule = &compile.Module{
+				Name: c.file,
+			}
+		}
+
+		fromModule, err := compile.Compile(c.file, compile.Filesystem(fsFrom))
+		if err != nil {
+			return pass, err
+		}
+		pass.Modules(fromModule, toModule)
+	}
+
+	// p will have lints as a field which we can sort in cli.
+
+	return pass, errs
+}
+
+// FS holds reference to components needed for git FS.
+type FS struct {
+	repo   *git.Repository
+	gitDir string
```

Comment:
naming: this is called git Dir but it's actually the repository root.
In git plumbing contexts, git dir refers to the ".git" directory.